### PR TITLE
Make ParsePatch more performant

### PR DESCRIPTION
**Pull request by bkcsoft**

### CODE REVIEW

**Comment by lafriks** [comment]
@bkcsoft build failure

File: models/git_diff.go [hunk 1]
```diff
@@ -247,30 +247,54 @@ func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader) (*D
...
-	isEOF := false
 	for !isEOF {
 		var linebuf bytes.Buffer
 		for {
-			b, err := input.ReadByte()
+			peek, err := input.Peek(maxLineCharacters)
+			if err != bufio.ErrBufferFull {
+				return nil, fmt.Errorf("PeekByte: %v", err)
+			}
+			newLine := bytes.Index(peek, []byte("\n"));
```
Comment (ethantkoenig) [inline models/git_diff.go]: Why not `bytes.IndexByte`?

File: models/git_diff.go [hunk 2]
```diff
@@ -247,30 +247,54 @@ func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader) (*D
...
 	)
 
-	input := bufio.NewReader(reader)
-	isEOF := false
 	for !isEOF {
 		var linebuf bytes.Buffer
 		for {
-			b, err := input.ReadByte()
+			peek, err := input.Peek(maxLineCharacters)
+			if err != bufio.ErrBufferFull {
```
Comment (ethantkoenig) [inline models/git_diff.go]: What if `err` is `nil`?

File: models/git_diff.go [hunk 3]
```diff
@@ -247,30 +247,54 @@ func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader) (*D
...
+				curFile.IsIncomplete = true
+				// We already know that we can read `len(peek)` amount of bytes,
+				//  hence no error-checking
+				input.Discard(len(peek))
+				continue
+			}
+			if curFile.IsIncomplete {
+				// Since we get here without hiting the above case, we've found a newline
+				//  and only discard that part.
+				input.Discard(newFile)
```
Comment (ethantkoenig) [inline models/git_diff.go]: `newLine`?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
Unit tests would be nice

**Review by sapk — COMMENTED** [review]
As it should provide performance improvement, benchmark would be nice to validate the result.

**Comment by bkcsoft** [comment]
@ethantkoenig Not sure what unit-tests you're referring to 🤔 

@sapk Should go without saying that reading chunks at a time is faster than reading it character by character 🙂

**Comment by ethantkoenig** [comment]
@bkcsoft I meant "It would be nice if you could add unit tests for the `ParsePatch` function" :smile:

**Comment by sapk** [comment]
@bkcsoft Just to make sure. I have done simple test and bench https://github.com/sapk-fork/gitea/commit/52dcc4674cc656d948ebfb3a50f618e99a069a43. 
It give a improvement of x2 but failed to pass simple test validation.

You can cherry-pick the commit if you want or I can put it in a separate PR.

**Comment by bkcsoft** [comment]
@sapk Thanks! 😄 I'll use that test to fix the issue 🙂

**Comment by lunny** [comment]
CI failed

**Comment by bkcsoft** [comment]
I'll leave this for the next week (at least) as I'll have limited time to work on it

**Comment by bkcsoft** [comment]
not working on this anymore, feel free to continue on it or not.

**Pull request closed without merging**